### DISCUSSION TITLE
Trouble getting service decoration to work

### DISCUSSION THREAD

**Discussion by gnito-org** [symfony/symfony#44279] [Q&A] [answered]
I'm trying to decorate Symfony\Component\Mailer\Mailer with App\Mailer\DecoratingMailer, using Symfony 6.0.0.-RC1.

`services.yaml` contains:

>     Symfony\Component\Mailer\Mailer: ~
>     App\Mailer\DecoratingMailer:
>       decorates: Symfony\Component\Mailer\Mailer

`symfony console debug:container` shows:

```
App\Mailer\DecoratingMailer  App\Mailer\DecoratingMailer
App\Mailer\DecoratingMailer.inner Symfony\Component\Mailer\Mailer
Symfony\Component\Mailer\Mailer alias for "App\Mailer\DecoratingMailer"
```
`DecoratingMailer.php` contains:
```
use Symfony\Component\Mailer\Mailer;

    public function __construct(Mailer $mailer) {
        $this->mailer = $mailer;
    }
```

`HomeController.php` includes:

```
    #[Route('/home', name: 'home')]
    public function index(Mailer $mailer): Response
    {
        return $this->render('home/index.html.twig', [
            'controller_name' => 'HomeController',
        ]);
    }
```

When I navigate to `127.0.0.1:8000/home`, it shows an error:

```
App\Controller\HomeController::index(): Argument #1 ($mailer) must be of type Symfony\Component\Mailer\Mailer, App\Mailer\DecoratingMailer given...
```

It feels as if I'm missing something basic, but I cannot figure out what it is.

Any help or advice would be greatly appreciated.

**Comment by gnito-org** [comment, marked as answer]
I got it to work, but to be honest, I don't really understand why it now works.

`services.yaml` contains:

```
    Symfony\Component\Mailer\MailerInterface:
        class: Symfony\Component\Mailer\Mailer
    App\Mailer\DecoratingMailer:
        decorates: Symfony\Component\Mailer\MailerInterface
```

`DecoratingMailer.php` contains:

```
use Symfony\Component\Mailer\MailerInterface;

class DecoratingMailer implements MailerInterface {

    public function __construct(MailerInterface $mailer) {
        $this->mailer = $mailer;
    }
```
`HomeController.php` includes:

```
    #[Route('/home', name: 'home')]
    public function index(MailerInterface $mailer): Response
    {
        return $this->render('home/index.html.twig', [
            'controller_name' => 'HomeController',
        ]);
    }
```

**Reply by mbabker** [reply]
You were typehinting the concrete `Mailer` class in your original example.  Your decorating class doesn't subclass the `Mailer` class, so the type error is correct.  Always typehint to an interface when one is available, generally, it's interfaces that let you swap out the concrete implementations in some way (like the decorator pattern).

**Comment by gnito-org** [comment]
Does the decorating class always need to implement the same interface as the decorated class?

**Reply by stof** [reply]
Well, the decorator needs to pass the typehint where it is used, so the answer is yes.

**Reply by stof** [reply]
if you don't implement the same type, what you do is not decoration anymore. It is only composition (and that's configured by injecting mailer into your own service using the normal feature)

**Comment by gnito-org** [comment]
I will submit a PR to expand the decoration description at https://symfony.com/doc/current/service_container/service_decoration.html.

It needs some elaboration around implementing interfaces and/or extending the decorated class.

**Reply by pablomartin89** [reply]
It's the PR submitted? I have the same 'TypeError' following the documentation. 

I don't know if it's required to implement interfaces,  but on the docs don't say anything about that.

**Reply by benblub** [reply]
After I also ran into the problem I created a gist. https://gist.github.com/benblub/b799c4d46facb7ff89244dede3cac02c
